### DISCUSSION TITLE
Creation of fixed size array (on the stack) with parameter for size

### DISCUSSION THREAD

**Discussion by jdonnerstag** [vlang/v#12698] [Ideas]
Fixed size arrays allocate memory on the stack, e.g. `ar := [10]int` . What if the size is not known at compile time, but a parameter passed to a function. It is still fixed size, but only known at runtime. How can I create a fixed size array *on the stack* with the size only known at runtime?

E.g. 
```
fn xyz(size int) {
   ar := [size]int
   ...
}
```
JalonSolov
 — 
heute um 19:16 Uhr
There are 2 ways to create a fixed size array.  [1,2,3]! for a fixed size array of the given literals, and [3]int for a fixed size array of 0s to start.  If you replace the 3 there with a variable, you'll create a fixed size array of whatever size that variable decrees.
Welp, never mind.  That should work, but doesn't.
In that case... you can't.  Since it requires a constant.
Unless you want to play games with a const and an unsafe block... <sigh>  It's entirely too easy to circumvent the rules.
Jdo
 — 
heute um 19:25 Uhr
Ok to create a feature request?
JalonSolov
 — 
heute um 19:27 Uhr
Personally, I think it's more of a bug, but sure.  An issue will at least make sure it's tracked.  🙂
Here's a question, though... why?
Why do you need a fixed size array whose size is only known at runtime?  If you include a good example of why it's useful, it might get done sooner.  🙂

My use case: I'm developing a V-lang based Rosie pattern matcher which has a small virtual machine. I'd like to make the runtime as fast as possible and already had good success with using a static array for the vm-stack. In tight loops there is significant difference between dynamic and static arrays. The runtime is using another dynamic array right now. It's size is not fixed (at compile time). A too large value wastes memory, and might still not be large enough for other use cases. So I thought, the user could provide a hint via a (config) parameter to the runtime function.

**Comment by Cons-Cat** [comment]
I think this is possible in C (using ``alloca()``, at least on Linux), so I imagine it could be done in V.

**Comment by Cons-Cat** [comment]
```v
#include <alloca.h>

fn C.alloca(int) &int

fn var_array(size int) {
	mut array := C.alloca(size)
	for i in 0 .. size {
		unsafe {
			array[i] = i
			println(array[i])
		}
	}
}

fn main() {
	var_array(5)
}
```
```
~/dump $ v run dump.v
0
1
2
3
4
```
Yeah, this basically works so it should be technologically feasible.

**Comment by mguinhos** [comment]
Alloca is not stable and should not be used freely, at least not outside in a unsafe block.
V should mark this kind of C function as unsafe

**Comment by mguinhos** [comment]
Alloca is more propense to unexpected behaviour

**Comment by Cons-Cat** [comment]
If V knows the size of its stack, then it can check for stack overflow in advance at this function's call site, and return optional so that it's safe.
